plugins/hls-tactics-plugin/src/Wingman/LanguageServer.hs

```diff
@@ -364,3 +379,62 @@ mkShowMessageParams ufm = ShowMessageParams (ufmSeverity ufm) $ T.pack $ show uf
 
 showLspMessage :: MonadLsp cfg m => ShowMessageParams -> m ()
 showLspMessage = sendNotification SWindowShowMessage
+
+
+-- This rule only exists for generating file diagnostics
+-- so the RuleResult is empty
+data WriteDiagnostics = WriteDiagnostics
+    deriving (Eq, Show, Typeable, Generic)
+
+instance Hashable WriteDiagnostics
+instance NFData   WriteDiagnostics
+instance Binary   WriteDiagnostics
+
+type instance RuleResult WriteDiagnostics = ()
+
+wingmanRules :: PluginId -> Rules ()
+wingmanRules plId = do
+  define $ \WriteDiagnostics nfp -> do
+    cfg <- flip configForPlugin plId <$> getClientConfigAction def
+    let severity = useProperty #hole_severity properties $ Just $ Plugin.plcConfig cfg
+    x <- use GetParsedModule nfp
+    case x of
+      Nothing ->
+        pure ([], Nothing)
+      Just pm -> do
+        let holes :: [Range]
+            holes =
+              everything (<>)
+                (mkQ mempty $ \case
+                  L span (HsVar _ (L _ name))
+                    | isHole (occName name) ->
+                        maybeToList $ srcSpanToRange span
+                  L span (HsUnboundVar _ (TrueExprHole occ))
+                    | isHole occ ->
+                        maybeToList $ srcSpanToRange span
+#if __GLASGOW_HASKELL__ <= 808
+                  L span (EWildPat _) ->
+                    maybeToList $ srcSpanToRange span
+#endif
+                  (_ :: LHsExpr GhcPs) -> mempty
+                ) $ pm_parsed_source pm
+        pure
+          ( fmap (\r -> (nfp, ShowDiag, mkDiagnostic severity r)) holes
+          , Just ()
+          )
+
+  action $ do
+    files <- getFilesOfInterest
+    void $ uses WriteDiagnostics $ Map.keys files
```

Comment:
Is this really necessary? It will be executed on every single keystroke, so please be very sure about it.

Reply:
Happy to get suggestions here; this is just cargo culted from hlint.

Reply:
What is the desired purpose? Do you want to refresh the code action diagnostics on every keystroke, or only when code actions are invoked?

Reply:
I must admit I'm in over my head here. I'd like a workflow that allows for refining holes by hand and via Wingman, that makes it easy to get to the next hole.

My original thought was to just automatically move to the next hole after running a code action, but LSP doesn't support that. And so if users need to do the movement themselves, it'd be nice if it worked under all circumstances.

Reply:
Have you seen this: https://github.com/haskell/ghcide/pull/889

The idea is that you can hook into GHC to record all the holes in the program while its being typechecked, no need to do a potentially slow SYB style traversal afterwards.

Reply:
>8.10 is a bit of a non-starter for me. If you're particularly concerned about perf, I'd be willing to write a custom SYB traversal that doesn't explore stupid parts of the tree, as well as making this diagnostic opt-in.

Reply:
Don't we already have diagnostics for holes? Would mapping over the diagnostics tagging the ones that look like holes work?

Reply:
I don't get diagnostics for holes! Do you?

Reply:
![2021-04-03-140400_grab](https://user-images.githubusercontent.com/4812498/113473193-a3869700-9485-11eb-953c-a2b2a20f77ba.png)

Reply:
I also get them.

Reply:
That's embarrassing. I had them disabled in my test project for some reason. Thanks. How can I get the existing diagnostics?

Reply:
You might want this:
https://github.com/haskell/haskell-language-server/blob/14b46e100c08877d1de9403084f56388f7b9cf56/ghcide/src/Development/IDE/Core/Shake.hs#L781-L784

Reply:
I'm confused about the flow here, and worried about races. 

`getDiagnostics` above requires an `IdeState`, which AFAICT isn't available inside of the shake `Action` where the new diagnostics need to be generated. Easy enough to work around this by mucking with the internals, but then `getDiagnostics` just reads a mutable variable. Without a proper rule that only fires when the diagnostics have changed, to ensure the other diagnostics get generated before I look at them, it seems like I'm just inviting race conditions.

But since diagnostics are generated (exclusively?) by rules, and that this new one will produce new diagnostics, it feels like this approach won't work. Unless I'm missing something obvious?

Reply:
No, you are right. The diagnostic store is mutable, and `getDiagnostics` will only read whatever is in there. 

But that said, I think you could add a dependency on the `Typecheck` rule, and then I believe `getDiagnostics` will find the `Typecheck` diagnostics in there.

Reply:
That makes sense, thanks!